src/hypothesis/strategies.py

```diff
@@ -1575,4 +1578,204 @@ def deferred(definition):
+            # \\?\UNC\[server]\[sharename]\
+            builds(lambda *x: tp().join(x),
+                   sep, sep, just(tp('?')), sep, just(tp('UNC')), sep, name,
+                   sep, name, sep),
+            # \\.\[physical_device]\
+            builds(lambda *x: tp().join(x),
+                   sep, sep, just(tp('.')), sep, name, sep),
+        ])
+
+        return draw(one_of(relative, drive, extended, network))
+
+
+@defines_strategy
+@composite
+def fspaths(draw, allow_pathlike=None, allow_existing=False):
+    """A strategy which generates filesystem path values.
+
+    The generated values include everything which the builtin
+    :func:`python:open` function accepts i.e. which won't lead to
+    :exc:`ValueError` or :exc:`TypeError` being raised.
+
+    Note that the range of the returned values depends on the operating
+    system, the Python version, and the filesystem encoding as returned by
+    :func:`sys.getfilesystemencoding`.
+
+    :param allow_pathlike:
+        By default makes the strategy also include objects implementing the
+        :class:`python:os.PathLike` interface when Python >= 3.6 is used. If
+        :obj:`python:False` no pathlike objects will be generated. If
+        :obj:`python:True` pathlike will be generated (Python >= 3.6 required)
+
+    :type allow_pathlike: :obj:`python:bool` or :obj:`python:None`
+
+    :param bool allow_existing:
+        If paths which exist on the filesystem should be generated.
+
+    .. versionadded:: 3.15
+
+    """
+    has_pathlike = hasattr(os, 'PathLike')
+
+    if allow_pathlike is None:
+        allow_pathlike = has_pathlike
+    if allow_pathlike and not has_pathlike:  # pragma: no cover
+        raise InvalidArgument(
+            'allow_pathlike: os.PathLike not supported, use None instead '
+            'to enable it only when available')
+
+    result_type = draw(sampled_from([bytes, text_type]))
+
+    def tp(s=''):
+        return _str_to_path(s, result_type)
+
+    special_component = sampled_from([tp(os.curdir), tp(os.pardir)])
+    normal_component = _filename(result_type)
+    path_component = one_of(normal_component, special_component)
+    extension = normal_component.map(lambda f: tp(os.extsep) + f)
+    root = _path_root(result_type)
+
+    def optional(st):
+        return one_of(st, just(result_type()))
+
+    sep = sampled_from([os.sep, os.altsep or os.sep]).map(tp)
+    path_part = builds(lambda s, l: s.join(l), sep, lists(path_component))
+    main_strategy = builds(lambda *x: tp().join(x),
+                           optional(root), path_part, optional(extension))
+
+    if not allow_existing:
+
+        def check_not_exists(path):
+            # os.path.exists returns False for broken symlinks, so use lstat
+            try:
+                os.lstat(path)
+            except OSError:  # pragma: no cover
+                return True
+            except ValueError:  # pragma: no cover
```

Comment:
> Yes. Too long paths seemed useful as they can happen when concatenating values in Python. But filtering is fine with me.

Hmm. I can see that there's a valid use case for technically valid file paths that can never exist. I'm not sure whether always filtering them or never filtering them is best, and am happy to leave that up to you, but I think it needs to be consistent which we do regardless of whether `allow_existing` is set.

> Would using os.lstat on Unix and os.path.exists on Windows help?

I think `os.path.exists` calls `os.stat` under the hood (which is equivalent to `os.lstat`), so it will be just as slow.

> os.lstat on Windows takes about 70ms per call here and slows everything down leading to hypothesis complaining.

Windows file system performance is generally quite bad, but that's *astonishingly* bad. Is that on Wine or on a real Windows?

I did some local testing and the performance I get is much better than that - more on the order of 20 microseconds. I tried the following to see if it would help:

```python
def fast_exists(f):
    try:
        with open(f, 'r'):
             return True
    except OSError:
        # TODO: Need to check if the failure was e.g. a permission error
        return False
```

On my system these were almost identically fast, but I wonder if the one you're having problems on is having to do some weird slow emulation of stat and file opening will be faster?